### ISSUE TITLE
[Feature] Expose Waiter API

### ISSUE THREAD

**Issue by fgirardey** [microsoft/playwright#6360]
TL;DR: Is it possible to expose your awesome [`Waiter`](https://github.com/microsoft/playwright/blob/master/src/client/waiter.ts) API?

I am currently working on a custom API to ease the development of e2e tests with `playwright` on our web app.

The idea is that the web app will emit some testing DOM events that could be listened by the e2e tests to get notified when something happened.

The problem is that the `page.waitForEvent()` function does not allow to listen custom DOM events so my first approach was to use `page.waitForFunction()` but it can't execute a custom predicate in the page context (which is required with the API i would like to design).

I don't really want to implement a polling algorithm myself and i would love to reuse your `Waiter.waitForPromise()` API so is there a plan to make it available?

**Comment by pavelfeldman** [comment]
This definitely sounds interesting!

**Comment by fgirardey** [comment]
Thank you for your time @pavelfeldman.

To give more specific details, for now I ended up with the following solution:

- Node side: Expose a `emitTestingEvent` function
- Browser side: The web app calls that function with a given testing event
- Node side: I push those events in a RxJS Subject

Then on e2e code I wait for a specific event with the following code:

```ts
await events$.pipe(
    filter((event) => predicate(event)),
    takeUntil(from(page.waitForTimeout(config.timeout))),
).toPromise()
```

So i can write testing scenarios like that:

```ts
await waitForTestingEvent(page, (event) => event.name === 'task:end');
```

I would rather like that playwright provide this capability to keep consistance with the other playwright APIs, something like:

```ts
await page.waitForEvent('testing-event', {
  predicate: (event) => event.name === 'task:end',
});
```

But it might be too opinionated for playwright so that's why i was thinking that exposing `Waiter` could be a good idea but actually `Waiter` does not give access to the event object in predicates…

I would love to hear your thoughts on this 😇

**Comment by rinogo** [comment]
I have the same problem. Essentially, I want to emit a custom event that I can listen for in our e2e Playwright tests. Without this functionality, I'm forced to use `waitForFunction()` along with some kind of a exposed/global variable which feels needlessly messy.

I, too, would love to be able to do something like:

```
await page.waitForEvent('my-custom-testing-event');
```

Is there a better alternative?

**Comment by pavelfeldman** [comment]
https://playwright.dev/docs/next/test-assertions#polling will ship in Playwright v1.21.

**Comment by fgirardey** [comment]
Amazing, Playwright has evolved so well recently, thank you for your hard work to make ours easier!
